server/src/main/java/org/elasticsearch/cluster/coordination/NodeJoinExecutor.java

```diff
@@ -355,6 +360,30 @@ private static void blockForbiddenVersions(TransportVersion joiningTransportVers
         }
     }
 
+    private Set<String> calculateAllAssumedClusterFeatures(DiscoveryNodes nodes, Map<String, Set<String>> nodeFeatures) {
```

Comment:
The javadocs for this should call out that it mutates the given `nodeFeatures` parameter.

Reply:
It operates on a copy
```
nodeFeatures = new HashMap<>(nodeFeatures);
```

Reply:
Fair point, though that's a shallow copy. If we also change the implementation to update the sets in-place, it's still mutating the contents of th emap.

Reply:
...which I guess is why [you said not to do it](https://github.com/elastic/elasticsearch/pull/118143/files/0939c46d6a2c5d9d5b47381888fa0ac012345152#r1882322576). 😓